common/client-core/src/client/statistics_control.rs

```diff
@@ -0,0 +1,174 @@
+const SNAPSHOT_INTERVAL_MS: u64 = 500;
+
+/// Launches and manages metrics collection and reporting.
+///
+/// This is designed to be generic to allow for multiple types of metrics to be collected and
+/// reported.
+pub(crate) struct StatisticsControl {
+    /// Keep store the different types of metrics collectors
+    stats: ClientStatsController,
+
+    /// Incoming packet stats events from other tasks
+    stats_rx: ClientStatsReceiver,
+
+    /// Channel to send stats report through the mixnet
+    report_tx: InputMessageSender,
+
+    /// Service-provider address to send stats reports
+    reporting_address: Recipient,
+}
+
+impl StatisticsControl {
+    pub(crate) fn new(
+        reporting_address: Recipient,
+        client_stats_id: String,
+        client_type: String,
+        report_tx: InputMessageSender,
+    ) -> (Self, ClientStatsSender) {
+        let (stats_tx, stats_rx) = tokio::sync::mpsc::unbounded_channel();
+
+        let stats = ClientStatsController::new(client_stats_id, client_type);
+
+        (
+            StatisticsControl {
+                stats,
+                stats_rx,
+                reporting_address,
+                report_tx,
+            },
+            ClientStatsSender::new(stats_tx),
+        )
+    }
+
+    async fn report_stats(&mut self) {
+        let stats_report = self.stats.build_report();
+
+        if let Ok(report_bytes) = stats_report.try_into() {
+            let report_message = InputMessage::new_regular(
+                self.reporting_address,
+                report_bytes,
+                TransmissionLane::General,
+                None,
+            );
+            if let Err(err) = self.report_tx.send(report_message).await {
+                log::error!("Failed to report client stats: {:?}", err);
+            } else {
+                self.stats.reset();
+            }
+        } else {
+            log::error!("Failed to serialize stats report. This should never happen");
+        }
+    }
+
+    pub(crate) async fn run_with_shutdown(&mut self, mut shutdown: nym_task::TaskClient) {
+        log::debug!("Started StatisticsControl with graceful shutdown support");
+
+        let report_interval = Duration::from_secs(STATS_REPORT_INTERVAL_SECS);
+        let mut report_interval = tokio::time::interval(report_interval);
+        let snapshot_interval = Duration::from_millis(SNAPSHOT_INTERVAL_MS);
+        let mut snapshot_interval = tokio::time::interval(snapshot_interval);
+
+        loop {
+            tokio::select! {
+                stats_event = self.stats_rx.recv() => match stats_event {
+                        Some(stats_event) => self.stats.handle_event(stats_event),
+                        None => {
+                            log::trace!("StatisticsControl: shutting down due to closed stats channel");
+                            break;
+                        }
+                },
+                _ = snapshot_interval.tick() => {
```

Comment:
note that `snapshot_interval` and `report_interval` will be run immediately upon starting which I'm not sure is what you want given there will be nothing to report

Reply:
Immediate report will at least be useful for the os_information so I see no problem for that, the snapshot stuff was already like that I think. it was just moved from elsewhere